src/pyclass.rs

```diff
@@ -107,134 +138,173 @@ pub trait PyClass:
     type BaseNativeType: PyTypeInfo + PyNativeType;
 }
 
-#[cfg(not(Py_LIMITED_API))]
-pub(crate) fn initialize_type_object<T>(
-    py: Python,
-    module_name: Option<&str>,
-    type_object: &mut ffi::PyTypeObject,
-) -> PyResult<()>
-where
-    T: PyClass,
-{
-    type_object.tp_doc = match T::DESCRIPTION {
-        // PyPy will segfault if passed only a nul terminator as `tp_doc`, ptr::null() is OK though.
-        "\0" => ptr::null(),
-        s if s.as_bytes().ends_with(b"\0") => s.as_ptr() as _,
-        // If the description is not null-terminated, create CString and leak it
-        s => CString::new(s)?.into_raw(),
-    };
+/// For collecting slot items.
+#[derive(Default)]
+pub(crate) struct TypeSlots(Vec<ffi::PyType_Slot>);
 
-    type_object.tp_base = T::BaseType::type_object_raw(py);
+impl TypeSlots {
+    fn push(&mut self, slot: c_int, pfunc: *mut c_void) {
+        self.0.push(ffi::PyType_Slot { slot, pfunc });
+    }
+    pub(crate) fn maybe_push(&mut self, slot: c_int, value: Option<*mut c_void>) {
+        if let Some(v) = value {
+            self.push(slot, v);
+        }
+    }
+}
 
-    type_object.tp_name = match module_name {
+fn tp_doc<T: PyClass>() -> PyResult<Option<*mut c_void>> {
```

Comment:
Can the return type of this just be `PyResult<*mut c_void>`, and use `std::ptr::null_mut()` instead of the `None` return?

Reply:
This way is easier for the caller.

Reply:
That's good enough for me!